ruby/phase-1.md

```diff
@@ -0,0 +1,575 @@
+  Ruby principles)
+- Backwards compatibility with Coinbase's Ruby SDK (current SDK customers will have to do a full
+  migration)
+- Reducing the number of different approaches to achieving the same goal (let the SDK user choose
+  what's best for their application/taste)
+
+### Tech stack && approach
+
+- Ruby version >= 2.7 (currently oldest fully supported version)
+- Type safety based on RBS or Sorbet (RBS will require Ruby version => 3.0)
+- Packaged as a Ruby gem (and published to [https://rubygems.org])
+- FFI or Rutie bridge into SDK Core (to be decided)
+
+
+## Key language considerations and restrictions
+
+We're assuming that the audience for this proposal might not be intimately familiar with Ruby. So
+before we jump into the main part here are some important considerations to keep in mind to avoid
+confusion during the discussion.
+
+*NOTE: The provided examples are not implying these are correct or incorrect ways of using Ruby, but
+simply demonstrate language specifics.*
+
+### Method invocations
+
+All parentheses in Ruby are optional (despite the presence or absence of method arguments). So these
+calls are identical:
+
+```ruby
+my_method()
+my_method
+```
+
+As well as these:
+
+```ruby
+my_method(a, b, c)
+my_method a, b, c
+```
+
+The most common style suggests not using parentheses when there are no arguments and using them
+otherwise. This rule is often ignored in a case of a DSL
+(e.g. [RSpec](https://relishapp.com/rspec/rspec-expectations/docs/built-in-matchers)).
+
+### Global scope
+
+Unlike other dynamic languages Ruby does not support requiring (or importing) files into a local
+scope. In practice that mean that everything that gets `require`d becomes globally available from
+every single part of the application. This applies to everything including constants, variables,
+methods, classes and modules. Here's an example:
+
+my_method.rb:
+
+```ruby
+def my_method(name)
+  puts "Hello World, #{name}!"
+end
+```
+
+my_class.rb:
+
+```ruby
+class MyClass
+  def run
+    my_method('Alice')
+  end
+end
+```
+
+main.rb:
+
+```ruby
+require './my_method.rb'
+require './my_class.rb'
+
+MyClass.new.run
+```
+
+Running `ruby main.rb` will print `Hello World, Alice!`.
+
+### "Monkey" patching
+
+In case of a variable or a method name collision the last definition overrides the previous one
+without hesitation. Same principle applies to methods defined on a class, but not the class itself —
+class definitions are merged together resulting in an altered class definition in the global scope.
+
+```ruby
+class MyClass
+  def my_method
+    puts 'Hello World!'
+  end
+end
+
+class MyClass
+  def my_method
+    puts 'Hello Monkeys!'
+  end
+
+  def new_method
+    puts 'Hello New World!'
+  end
+end
+
+my = MyClass.new
+my.my_method
+# => Hello Monkeys!
+my.new_method
+# => Hello New World
+```
+
+### Concurrency
+
+Ruby can be a tricky language for running concurrent applications. It supports kernel threads,
```

Comment:
To be clear, the Ruby language doesn't impose a GIL. The CRuby runtime has a GIL but alternative runtimes like JRuby and TruffleRuby have no GIL at all. However, CRuby is the most prevalent runtime.